File: app/src/main/java/jp/linanfine/dsma/util/auth/GoogleAuthManager.java
package jp.linanfine.dsma.util.auth;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import jp.linanfine.dsma.BuildConfig;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class GoogleAuthManager {
    public static final int RC_SIGN_IN = 9001;

    private static final String TAG = "GoogleAuthManager";
    private static final String BASE_URL = "https://fnapi.fia-s.com/api";
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    private static GoogleAuthManager instance;
    private final OkHttpClient client;
    private GoogleSignInClient googleSignInClient;
    private GoogleAuthCallback callback;

    // シングルトンパターン
    public static synchronized GoogleAuthManager getInstance() {
        if (instance == null) {
            instance = new GoogleAuthManager();
        }
        return instance;
    }

    private GoogleAuthManager() {
        client = new OkHttpClient();
    }

    // Googleサインインクライアントの初期化
    public void initialize(Context context) {
        Log.d(TAG, "Using Google Client ID: " + BuildConfig.GOOGLE_CLIENT_ID);
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(BuildConfig.GOOGLE_CLIENT_ID)
                .requestEmail()
                .build();

        googleSignInClient = GoogleSignIn.getClient(context, gso);
    }

    // Googleサインインを開始
    public void signIn(Activity activity, GoogleAuthCallback callback) {
        this.callback = callback;
        Log.d(TAG, "Starting Google Sign In");
        Intent signInIntent = googleSignInClient.getSignInIntent();
        Log.d(TAG, "Created sign in intent: " + signInIntent);
        activity.startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    public void handleSignInResult(Intent data) {
        try {
            Log.d(TAG, "Handling sign in result...");
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);

            // 追加デバッグログ（TAGを使用）
            if (task.isSuccessful()) {
                Log.d(TAG, "Task is successful");
            } else {
                Exception e = task.getException();
                Log.e(TAG, "Task failed with exception: " + (e != null ? e.getMessage() : "null"));
            }

            GoogleSignInAccount account = task.getResult(ApiException.class);
            if (account != null && account.getIdToken() != null) {
                Log.d(TAG, "Sign in success, id token: " + account.getIdToken().substring(0, 10) + "...");
                callback.onSuccess(account.getIdToken(), account);
            } else {
                Log.w(TAG, "Sign in failed: account or token is null");
                callback.onFailure("アカウント情報の取得に失敗しました");
            }
        } catch (ApiException e) {
            Log.w(TAG, "Google sign in failed: " + e.getStatusCode() + ", " + e.getMessage(), e);
            callback.onFailure("Google認証に失敗しました: " + e.getStatusCode());
        } catch (Exception e) {
            Log.e(TAG, "Unexpected error in sign in handling", e);
            callback.onFailure("認証処理中に予期せぬエラーが発生しました: " + e.getMessage());
        }
    }

    // Googleアカウントからユーザーを検索
    public void findPlayerByGoogleToken(String idToken, final FindPlayerCallback callback) {
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("googleToken", idToken);
        } catch (JSONException e) {
            callback.onFailure("JSONの作成に失敗しました: " + e.getMessage());
            return;
        }

        RequestBody body = RequestBody.create(JSON, jsonBody.toString());
        Request request = new Request.Builder()
                .url(BASE_URL + "/auth/find-player-by-google")
                .post(body)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.onFailure("ネットワークエラー: " + e.getMessage());
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (!response.isSuccessful()) {
                    callback.onFailure("サーバーエラー: " + response.code());
                    return;
                }

                String responseBody = response.body() != null ? response.body().string() : "{}";
                try {
                    JSONObject jsonResponse = new JSONObject(responseBody);
                    boolean success = jsonResponse.optBoolean("success", false);
                    boolean found = jsonResponse.optBoolean("found", false);
                    String message = "";

                    if (success && found) {
                        JSONObject playerObj = jsonResponse.getJSONObject("player");
                        String id = playerObj.getString("id");
                        String name = playerObj.getString("name");
                        Player player = new Player(id, name);
                        message = "Googleアカウントで認証され、ユーザー「" + name + "」としてログインしました";
                        callback.onSuccess(true, found, player, message);
                    } else if (success) {
                        message = "Googleアカウントに紐づけられたユーザーが見つかりませんでした。新規ユーザー登録を行ってください。";
                        callback.onSuccess(true, false, null, message);
                    } else {
                        String error = jsonResponse.optString("error", "不明なエラー");
                        callback.onFailure(error);
                    }
                } catch (JSONException e) {
                    callback.onFailure("レスポンスの解析に失敗しました: " + e.getMessage());
                }
            }
        });
    }

    // ユーザーとGoogleアカウントを紐づける
    public void connectUserWithGoogleToken(String playerId, String idToken, final ConnectGoogleCallback callback) {
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("playerId", playerId);
            jsonBody.put("googleToken", idToken);
        } catch (JSONException e) {
            callback.onFailure("JSONの作成に失敗しました: " + e.getMessage());
            return;
        }

        RequestBody body = RequestBody.create(JSON, jsonBody.toString());
        Request request = new Request.Builder()
                .url(BASE_URL + "/auth/connect-google")
                .post(body)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.onFailure("ネットワークエラー: " + e.getMessage());
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (!response.isSuccessful()) {
                    callback.onFailure("サーバーエラー: " + response.code());
                    return;
                }

                String responseBody = response.body() != null ? response.body().string() : "{}";
                try {
                    JSONObject jsonResponse = new JSONObject(responseBody);
                    boolean success = jsonResponse.optBoolean("success", false);

                    if (success) {
                        Player player = null;
                        if (jsonResponse.has("player")) {
                            JSONObject playerObj = jsonResponse.getJSONObject("player");
                            String id = playerObj.getString("id");
                            String name = playerObj.getString("name");
                            player = new Player(id, name);
                        }
                        String message = "Googleアカウントとの連携が完了しました";
                        callback.onSuccess(player, message);
                    } else {
                        String error = jsonResponse.optString("error", "不明なエラー");
                        callback.onFailure(error);
                    }
                } catch (JSONException e) {
                    callback.onFailure("レスポンスの解析に失敗しました: " + e.getMessage());
                }
            }
        });
    }

    // Googleアカウントとの連携を解除
    public void unlinkGoogleFromPlayer(String playerId, final UnlinkGoogleCallback callback) {
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("playerId", playerId);
        } catch (JSONException e) {
            callback.onFailure("JSONの作成に失敗しました: " + e.getMessage());
            return;
        }

        RequestBody body = RequestBody.create(JSON, jsonBody.toString());
        Request request = new Request.Builder()
                .url(BASE_URL + "/auth/unlink-google")
                .post(body)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.onFailure("ネットワークエラー: " + e.getMessage());
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (!response.isSuccessful()) {
                    callback.onFailure("サーバーエラー: " + response.code());
                    return;
                }

                String responseBody = response.body() != null ? response.body().string() : "{}";
                try {
                    JSONObject jsonResponse = new JSONObject(responseBody);
                    boolean success = jsonResponse.optBoolean("success", false);

                    if (success) {
                        String message = jsonResponse.optString("message", "Googleアカウントとの連携を解除しました");
                        callback.onSuccess(message);
                    } else {
                        String error = jsonResponse.optString("error", "不明なエラー");
                        callback.onFailure(error);
                    }
                } catch (JSONException e) {
                    callback.onFailure("レスポンスの解析に失敗しました: " + e.getMessage());
                }
            }
        });
    }

    // ユーザーの存在確認とGoogle連携状態を検証
    public void validateUser(String userId, final ValidateUserCallback callback) {
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("id", userId);
        } catch (JSONException e) {
            callback.onFailure("JSONの作成に失敗しました: " + e.getMessage());
            return;
        }

        RequestBody body = RequestBody.create(JSON, jsonBody.toString());
        Request request = new Request.Builder()
                .url(BASE_URL + "/validate-user")
                .post(body)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.onFailure("ネットワークエラー: " + e.getMessage());
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (!response.isSuccessful()) {
                    callback.onFailure("サーバーエラー: " + response.code());
                    return;
                }

                String responseBody = response.body() != null ? response.body().string() : "{}";
                try {
                    JSONObject jsonResponse = new JSONObject(responseBody);
                    boolean exists = jsonResponse.optBoolean("exists", false);
                    boolean isGoogleLinked = jsonResponse.optBoolean("isGoogleLinked", false);

                    callback.onSuccess(exists, isGoogleLinked, null);
                } catch (JSONException e) {
                    callback.onFailure("レスポンスの解析に失敗しました: " + e.getMessage());
                }
            }
        });
    }

    // Google連携情報の保存
    public void saveGoogleLinkStatus(Context context, String playerId, boolean isLinked) {
        SharedPreferences prefs = context.getSharedPreferences("FlareNotePrefs", Context.MODE_PRIVATE);
        prefs.edit().putBoolean("isGoogleLinked_" + playerId, isLinked).apply();
    }

    // Google連携情報の取得
    public boolean getGoogleLinkStatus(Context context, String playerId) {
        SharedPreferences prefs = context.getSharedPreferences("FlareNotePrefs", Context.MODE_PRIVATE);
        return prefs.getBoolean("isGoogleLinked_" + playerId, false);
    }

    // コールバックインターフェース
    public interface GoogleAuthCallback {
        void onSuccess(String idToken, GoogleSignInAccount account);
        void onFailure(String error);
    }

    public interface FindPlayerCallback {
        void onSuccess(boolean success, boolean found, Player player, String message);
        void onFailure(String error);
    }

    public interface ConnectGoogleCallback {
        void onSuccess(Player player, String message);
        void onFailure(String error);
    }

    public interface UnlinkGoogleCallback {
        void onSuccess(String message);
        void onFailure(String error);
    }

    public interface ValidateUserCallback {
        void onSuccess(boolean exists, boolean isGoogleLinked, String message);
        void onFailure(String error);
    }

    // Playerモデルクラス
    public static class Player {
        private final String id;
        private final String name;

        public Player(String id, String name) {
            this.id = id;
            this.name = name;
        }

        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }
    }
}

File: app/src/main/kotlin/jp/linanfine/dsma/util/auth/GoogleAuthManager.kt
package jp.linanfine.dsma.util.auth

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.credentials.CredentialManager
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import androidx.credentials.exceptions.GetCredentialException
import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.android.libraries.identity.googleid.GoogleIdTokenParsingException
import jp.linanfine.dsma.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import androidx.core.content.edit

class GoogleAuthManager private constructor() {
    companion object {
        private const val TAG = "GoogleAuthManager"
        private const val BASE_URL = "https://fnapi.fia-s.com/api"
        private val JSON = "application/json; charset=utf-8".toMediaType()

        @Volatile
        private var instance: GoogleAuthManager? = null

        fun getInstance(): GoogleAuthManager {
            return instance ?: synchronized(this) {
                instance ?: GoogleAuthManager().also { instance = it }
            }
        }
    }

    private val httpClient = OkHttpClient()
    private lateinit var prefs: SharedPreferences

    fun init(context: Context) {
        prefs = context.getSharedPreferences("FlareNotePrefs", Context.MODE_PRIVATE)
    }

    /**
     * Google認証を実行してIDトークンを取得
     */
    private suspend fun authenticate(context: Context): String {
        try {
            val signInWithGoogleOption = GetSignInWithGoogleOption
                .Builder(BuildConfig.GOOGLE_CLIENT_ID)
                .build()

            val credentialManager = CredentialManager.create(context)
            val request = GetCredentialRequest.Builder()
                .addCredentialOption(signInWithGoogleOption)
                .build()

            val result = credentialManager.getCredential(
                request = request,
                context = context
            )

            val credential = result.credential

            return when (credential) {
                is CustomCredential -> {
                    if (credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
                        try {
                            val googleIdTokenCredential = GoogleIdTokenCredential
                                .createFrom(credential.data)
                            googleIdTokenCredential.idToken
                        } catch (e: GoogleIdTokenParsingException) {
                            Log.e(TAG, "IDトークン解析エラー", e)
                            throw IOException("Googleトークン解析エラー: ${e.message}")
                        }
                    } else {
                        Log.e(TAG, "不明な認証情報タイプ: ${credential.type}")
                        throw IOException("不明な認証情報タイプ")
                    }
                }
                else -> {
                    Log.e(TAG, "想定外の認証情報: ${credential.javaClass.simpleName}")
                    throw IOException("想定外の認証情報タイプ")
                }
            }
        } catch (e: GetCredentialException) {
            Log.e(TAG, "Google認証エラー", e)
            throw IOException("Google認証エラー: ${e.message}")
        }
    }

    /**
     * Google認証でユーザーを検索する
     */
    private suspend fun findUserWithGoogle(context: Context): FindUserResult {
        val idToken = authenticate(context)
        return findUserByGoogleToken(idToken)
    }

    fun findUserWithGoogleSync(context: Context): FindUserResult {
        return runBlocking {
            findUserWithGoogle(context)
        }
    }

    /**
     * 既存ユーザーとGoogleアカウントを連携する
     */
    private suspend fun connectWithGoogle(context: Context, userId: String): User {
        val idToken = authenticate(context)
        return connectUserWithGoogle(userId, idToken)
    }

    fun connectWithGoogleSync(context: Context, userId: String): User {
        return runBlocking {
            connectWithGoogle(context, userId)
        }
    }

    /**
     * Googleアカウントからユーザーを検索
     */
    private suspend fun findUserByGoogleToken(idToken: String): FindUserResult = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("googleToken", idToken)
            }

            val request = Request.Builder()
                .url("$BASE_URL/auth/find-player-by-google")
                .post(requestBody.toString().toRequestBody(JSON))
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("サーバーエラー: ${response.code}")
                }

                val responseBody = response.body?.string() ?: throw IOException("レスポンスボディが空です")
                val json = JSONObject(responseBody)

                val result = FindUserResult().apply {
                    success = json.optBoolean("success", false)
                    found = json.optBoolean("found", false)

                    if (success && found) {
                        val playerJson = json.getJSONObject("player")
                        player = User(
                            playerJson.getString("id"),
                            playerJson.getString("name")
                        )
                    }
                }

                result
            }
        } catch (e: Exception) {
            when (e) {
                is JSONException -> throw IOException("リクエストまたはレスポンスの解析に失敗しました: ${e.message}")
                else -> throw e
            }
        }
    }

    /**
     * ユーザーとGoogleアカウントを紐づける
     */
    private suspend fun connectUserWithGoogle(userId: String, idToken: String): User = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("playerId", userId)
                put("googleToken", idToken)
            }

            val request = Request.Builder()
                .url("$BASE_URL/auth/connect-google")
                .post(requestBody.toString().toRequestBody(JSON))
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("サーバーエラー: ${response.code}")
                }

                val responseBody = response.body?.string() ?: throw IOException("レスポンスボディが空です")
                val json = JSONObject(responseBody)

                val success = json.optBoolean("success", false)
                if (success && json.has("player")) {
                    val playerJson = json.getJSONObject("player")
                    User(
                        playerJson.getString("id"),
                        playerJson.getString("name")
                    )
                } else {
                    val error = json.optString("error", "不明なエラー")
                    throw IOException(error)
                }
            }
        } catch (e: Exception) {
            when (e) {
                is JSONException -> throw IOException("リクエストまたはレスポンスの解析に失敗しました: ${e.message}")
                else -> throw e
            }
        }
    }

    /**
     * Googleアカウントとの紐づけを解除
     */
    private suspend fun disconnectGoogle(userId: String): String = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("playerId", userId)
            }

            val request = Request.Builder()
                .url("$BASE_URL/auth/unlink-google")
                .post(requestBody.toString().toRequestBody(JSON))
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("サーバーエラー: ${response.code}")
                }

                val responseBody = response.body?.string() ?: throw IOException("レスポンスボディが空です")
                val json = JSONObject(responseBody)

                val success = json.optBoolean("success", false)
                if (success) {
                    json.optString("message", "Google連携を解除しました")
                } else {
                    val error = json.optString("error", "不明なエラー")
                    throw IOException(error)
                }
            }
        } catch (e: Exception) {
            when (e) {
                is JSONException -> throw IOException("リクエストまたはレスポンスの解析に失敗しました: ${e.message}")
                else -> throw e
            }
        }
    }

    fun disconnectGoogleSync(userId: String): String {
        return runBlocking {
            disconnectGoogle(userId)
        }
    }

    /**
     * サーバー上のユーザー状態を検証
     */
    private suspend fun validateUser(userId: String): UserValidationResult = withContext(Dispatchers.IO) {
        try {
            val requestBody = JSONObject().apply {
                put("id", userId)
            }

            val request = Request.Builder()
                .url("$BASE_URL/validate-user")
                .post(requestBody.toString().toRequestBody(JSON))
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("サーバーエラー: ${response.code}")
                }

                val responseBody = response.body?.string() ?: throw IOException("レスポンスボディが空です")
                val json = JSONObject(responseBody)

                UserValidationResult().apply {
                    exists = json.optBoolean("exists", false)
                    isGoogleLinked = json.optBoolean("isGoogleLinked", false)
                }
            }
        } catch (e: Exception) {
            when (e) {
                is JSONException -> throw IOException("リクエストまたはレスポンスの解析に失敗しました: ${e.message}")
                else -> throw e
            }
        }
    }

    fun validateUserSync(userId: String): UserValidationResult {
        return runBlocking {
            validateUser(userId)
        }
    }

    /**
     * Google連携状態をローカルに保存
     */
    fun saveGoogleLinkStatus(userId: String, isLinked: Boolean) {
        prefs.edit { putBoolean("google_linked_$userId", isLinked) }
    }

    /**
     * Google連携状態をローカルから取得
     */
    fun getGoogleLinkStatus(userId: String): Boolean {
        return prefs.getBoolean("google_linked_$userId", false)
    }

    // 内部クラス定義
    data class User(val id: String, val name: String)

    class FindUserResult {
        var success: Boolean = false
        var found: Boolean = false
        var player: User? = null
    }

    class UserValidationResult {
        var exists: Boolean = false
        var isGoogleLinked: Boolean = false
    }
}